### ISSUE TITLE
Metadata schemas causing slow replication

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/tskit#1495]
When I run `ts.allele_frequency_spectrum` on a one-tree TS produced by `sim_ancestry`, it's almost an order of magnitude slower than when I run it on the tree sequence produced by the old `simulate` command:

```python
from time import time

import msprime
import numpy as np

T = 0.4
B=4
numrep=50000
nsamp=4
t = time()
demographic_events=[msprime.InstantaneousBottleneck(time=T, strength=B, population=0)]
reps1 = msprime.simulate(sample_size=nsamp, Ne=1/2, num_replicates=numrep, demographic_events=demographic_events, random_seed=1234)
print(f"msprime.simulate: {time()-t}s")

t = time()
demography = msprime.Demography()
demography.add_population(initial_size=1/2)
demography.add_instantaneous_bottleneck(time=T, strength=B, population=0)
reps2 = msprime.sim_ancestry(
    samples=nsamp, ploidy=1, num_replicates=numrep, demography=demography, random_seed=1234)
print(f"msprime.sim_ancestry: {time()-t}s")

t = time()
Blist1=np.zeros((numrep, nsamp+1))
for rep_index, ts in enumerate(reps1):
    afs=ts.allele_frequency_spectrum(mode="branch", polarised=True, span_normalise=False)
    Blist1[rep_index]+= afs
print(f"msprime.simulate afs calc: {time()-t}s")

t = time()
Blist2=np.zeros((numrep, nsamp+1))
for rep_index, ts in enumerate(reps2):
    afs=ts.allele_frequency_spectrum(mode="branch", polarised=True, span_normalise=False)
    Blist2[rep_index]+= afs
print(f"msprime.sim_ancestry afs calc: {time()-t}s")
``` 

gives:

```
msprime.simulate: 0.00606989860534668s
msprime.sim_ancestry: 0.012331008911132812s
msprime.simulate afs calc: 19.11087679862976s
msprime.sim_ancestry afs calc: 130.13320803642273s
```

This is causing problems with timeouts when running the new bottlenecks tutorial from Konrad. Am I misunderstanding the initial_size argument or something?

**Comment by hyanwong** [comment]
Following up on this, I think `demography.add_instantaneous_bottleneck` must be doing something different from `msprime.InstantaneousBottleneck`, as the AFS results are different in the two examples above.

**Comment by jeromekelleher** [comment]
This doesn't have anything to do with AFS, it's the actual simulation on the two code paths that's taking the time (just comment out the AFS calls as an easy way to chekc). I think the problem is something to do with metadata, because we have Schemas  in one and not the other.

Can you check that the bottlenecks are equivalent though by looking at the mean TMRCA in both cases, please?

**Comment by jeromekelleher** [comment]
Also, I think this is an msprime bug, not tskit?

**Comment by benjeffery** [comment]
This is a tskit performance issue. 90% of the time is spent parsing and validating the metadata schema, twice per replicate in `ancestry.py:1447` (`TableCollection.fromdict`) and `ancestry.py:1463` (`TableCollection.tree_sequence`)

This should be an easy fix.

**Comment by benjeffery** [comment]
Cool, a one line change has got it to:
```
msprime.simulate: 0.0020346641540527344s
msprime.sim_ancestry: 0.0031843185424804688s
msprime.simulate afs calc: 6.3704001903533936s
msprime.sim_ancestry afs calc: 6.574406147003174s
```
Will PR.

**Comment by jeromekelleher** [comment]
Excellent!

**Comment by hyanwong** [comment]
> This doesn't have anything to do with AFS, it's the actual simulation on the two code paths that's taking the time (just comment out the AFS calls as an easy way to chekc). 

Oh, of course. I'm thinking that the call to simulate with `num_replicates` actually runs the simulation, but it's just a generator. Sorry.
 
> Can you check that the bottlenecks are equivalent though by looking at the mean TMRCA in both cases, please?

The expected SFS seems different:

```
from time import time

import matplotlib.pyplot as plt
import msprime
import numpy as np

T = 0.4
B=4
numrep=50000
nsamp=4
demographic_events=[msprime.InstantaneousBottleneck(time=T, strength=B, population=0)]
reps1 = msprime.simulate(sample_size=nsamp, Ne=1/2, num_replicates=numrep, demographic_events=demographic_events, random_seed=1234)

demography = msprime.Demography()
demography.add_population(initial_size=1/2)
demography.add_instantaneous_bottleneck(time=T, strength=B, population=0)
reps2 = msprime.sim_ancestry(
    samples=nsamp, ploidy=1, num_replicates=numrep, demography=demography, random_seed=1234)

t = time()
Blist1=np.zeros((numrep, nsamp+1))
for rep_index, ts in enumerate(reps1):
    afs=ts.allele_frequency_spectrum(mode="branch", polarised=True, span_normalise=False)
    Blist1[rep_index]+= afs
print(f"msprime.simulate: {time()-t}s")

t = time()
Blist2=np.zeros((numrep, nsamp+1))
for rep_index, ts in enumerate(reps2):
    afs=ts.allele_frequency_spectrum(mode="branch", polarised=True, span_normalise=False)
    Blist2[rep_index]+= afs
print(f"msprime.sim_ancestry: {time()-t}s")


s=1-np.exp(-B)
p=s*(-6 + 15*s - 20 * np.power(s,2) + 15 * np.power(s,3) - 6 * np.power(s,4) + np.power(s,5))

expsfsBottlN= [2/15*(np.exp(-6*T)*(15 *np.exp(6*T) - 9 *np.exp(5*T)*s - 
    5*np.exp(3*T)*s*(3 - 3*s + np.power(s,2)) + p)),
    1/5*np.exp(-6*T)*(5*np.exp(6*T) - 6*np.exp(5*T)*s - p),
    2/15*np.exp(-6*T)*(5*np.exp(6*T) - 9*np.exp(5*T)*s + 5*np.exp(3*T)*s*(3-3*s + np.power(s,2))+ p)]

expsfsBottlN/=np.sum(expsfsBottlN)

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))
index = np.arange(1,4)
bar_width = 0.4
opacity = 0.9

data=np.mean(Blist1, axis=0)
data/=np.sum(data)

expsfs = ax1.bar(index+ bar_width, expsfsBottlN, bar_width, alpha=opacity, label='exp')
simsfs = ax1.bar(index+ 2*bar_width, data[1:4], bar_width, alpha=opacity, label='exp')
ax1.set_title("msprime.InstantaneousBottleneck")

data=np.mean(Blist2, axis=0)
data/=np.sum(data)

expsfs = ax2.bar(index+ bar_width, expsfsBottlN, bar_width, alpha=opacity, label='exp')
simsfs = ax2.bar(index+ 2*bar_width, data[1:4], bar_width, alpha=opacity, label='exp')
ax2.set_title("demography.add_instantaneous_bottleneck")
```

![image](https://user-images.githubusercontent.com/4699014/122059302-65421800-cde4-11eb-9c2e-e0c8afee1701.png)

**Comment by hyanwong** [comment]
And here are the mean tMRCAs:

```
from time import time

import matplotlib.pyplot as plt
import msprime
import numpy as np

T = 0.4
B=4
numrep=50000
nsamp=4
demographic_events=[msprime.InstantaneousBottleneck(time=T, strength=B, population=0)]
reps1 = msprime.simulate(sample_size=nsamp, Ne=1/2, num_replicates=numrep, demographic_events=demographic_events, random_seed=1234)

demography = msprime.Demography()
demography.add_population(initial_size=1/2)
demography.add_instantaneous_bottleneck(time=T, strength=B, population=0)
reps2 = msprime.sim_ancestry(
    samples=nsamp, ploidy=1, num_replicates=numrep, demography=demography, random_seed=1234)

t = time()
tMRCAs = np.zeros(numrep)
Blist1=np.zeros((numrep, nsamp+1))
for rep_index, ts in enumerate(reps1):
    afs=ts.allele_frequency_spectrum(mode="branch", polarised=True, span_normalise=False)
    tMRCA = 0
    for tree in ts.trees():
        tMRCA += ts.node(tree.root).time * tree.interval.span
    tMRCAs[rep_index] = tMRCA/ts.sequence_length
    Blist1[rep_index]+= afs
print(f"msprime.simulate: {time()-t}s: mean tMRCA = {np.mean(tMRCAs)}")

t = time()
Blist2=np.zeros((numrep, nsamp+1))
for rep_index, ts in enumerate(reps2):
    afs=ts.allele_frequency_spectrum(mode="branch", polarised=True, span_normalise=False)
    tMRCA = 0
    for tree in ts.trees():
        tMRCA += ts.node(tree.root).time * tree.interval.span
    tMRCAs[rep_index] = tMRCA/ts.sequence_length
    Blist2[rep_index]+= afs
print(f"msprime.sim_ancestry: {time()-t}s: mean tMRCA = {np.mean(tMRCAs)}")
```

giving:

```
msprime.simulate: 27.158241033554077s: mean tMRCA = 0.41306112646877324
msprime.sim_ancestry: 167.79602098464966s: mean tMRCA = 0.3603965555359294
```

Should I open this as an msprime issue?

**Comment by jeromekelleher** [comment]
> Should I open this as an msprime issue?

Yes please - probably a factor of two gone awry here.

**Comment by hyanwong** [comment]
Fixed in #1496
